File: 07-Builder/src/cpp/main.cpp
#include <iostream>
#include "HouseBuilder.h"
#include "HouseDirector.h"

int main() {
    std::cout << "=== Builder Pattern Demo ===\n\n";

    // 1. Build a custom house step by step
    House custom_house = HouseBuilder()
            .rooms(4)
            .floors(2)
            .garage()
            .pool()
            .roof_type("tile")
            .build();
    std::cout << "Custom:  " << custom_house.to_string() << "\n";

    // 2. Use the Director for pre-defined configurations
    HouseDirector director;

    House luxury = director.build_luxury_house(HouseBuilder());
    std::cout << "Luxury:  " << luxury.to_string() << "\n";

    House simple = director.build_simple_house(HouseBuilder());
    std::cout << "Simple:  " << simple.to_string() << "\n";

    House family = director.build_family_house(HouseBuilder());
    std::cout << "Family:  " << family.to_string() << "\n";

    // 3. Minimal house — only defaults
    House minimal = HouseBuilder().build();
    std::cout << "Minimal: " << minimal.to_string() << "\n";

    return 0;
}

File: 07-Builder/src/rust/main.rs
mod house;
mod director;

use house::HouseBuilder;
use director::HouseDirector;

fn main() {
    println!("=== Builder Pattern Demo ===\n");

    // 1. Build a custom house step by step
    let custom_house = HouseBuilder::new()
        .rooms(4)
        .floors(2)
        .garage()
        .pool()
        .roof_type("tile")
        .build();
    println!("Custom:  {}", custom_house);

    // 2. Use the Director for pre-defined configurations
    let luxury = HouseDirector::build_luxury_house();
    println!("Luxury:  {}", luxury);

    let simple = HouseDirector::build_simple_house();
    println!("Simple:  {}", simple);

    let family = HouseDirector::build_family_house();
    println!("Family:  {}", family);

    // 3. Minimal house — only defaults
    let minimal = HouseBuilder::new().build();
    println!("Minimal: {}", minimal);
}
